contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,364 @@
+
+* Goal 2: Implement CSI volume snapshot support.
+An external snapshot controller will be deployed with other external components (e.g., external-attacher, external-provisioner) for each CSI Volume Driver.
+
+* Goal 3: Provide a convenient way of creating new and restoring existing volumes from snapshots.
+
+
+### Non-Goals
+
+The following are non-goals for the current phase, but will be considered at a later phase.
+
+* Goal 4: Offer application-consistent snapshots by providing pre/post snapshot hooks to freeze/unfreeze applications and/or unmount/mount file system.
+
+* Goal 5: Provide higher-level management, such as backing up and restoring a pod and statefulSet, and creating a consistent group of snapshots.
+
+
+## Design Details
+
+In this proposal, volume snapshots are considered as another type of storage resources managed by Kubernetes. Therefore the snapshot API and controller follow the design pattern of existing volume management. There are three APIs, VolumeSnapshot and VolumeSnapshotContent, and VolumeSnapshotClass which are similar to the structure of PersistentVolumeClaim and PersistentVolume, and storageClass. The external snapshot controller functions similar to the in-tree PV controller. With the snapshots APIs, we also propose to add a new data source struct in PersistentVolumeClaim (PVC) API in order to support restore snapshots to volumes. The following section explains in more details about the APIs and the controller design.
+
+
+### Snapshot API Design
+
+The API design of VolumeSnapshot and VolumeSnapshotContent is modeled after PersistentVolumeClaim and PersistentVolume. In the first version, the VolumeSnapshot lifecycle is completely independent of its volumes source (PVC). When PVC/PV is deleted, the corresponding VolumeSnapshot and VolumeSnapshotContents objects will continue to exist.
+
+#### The `VolumeSnapshot` Object
+
+```GO
+
+// The volume snapshot object accessible to the user. Upon successful creation of the actual
+// snapshot by the volume provider it is bound to the corresponding VolumeSnapshotContent through
+// the VolumeSnapshotSpec
+type VolumeSnapshot struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec represents the desired state of the snapshot
+	Spec VolumeSnapshotSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status represents the latest observer state of the snapshot
+	// +optional
+	Status VolumeSnapshotStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+type VolumeSnapshotList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Items           []VolumeSnapshot `json:"items"`
+}
+
+// VolumeSnapshotSpec is the desired state of the volume snapshot
+type VolumeSnapshotSpec struct {
+        // PersistentVolumeClaimName is the name of the PVC being snapshotted
```

Comment:
volumesnapshot creations supports two types: dynamically creation of volumesnapshot and statically binding of snapshot and snapshotContent. In case of dynamically creation, PVCName must be provided. For statically binding, PVC can be nil, and user has to create VolumeSnapshotContent and bind it with VolumeSnapshot